### DISCUSSION TITLE
Is it possible to use this API inside the cloudflare workers framework/engine?

### DISCUSSION THREAD

**Discussion by null-prophet** [discordjs/discord.js#8428] [Q&A] [answered]
I came across this amazing API and wondered if it was compatible with the cloudflare worker model.

I've seen an example from discord setting up a worker to respond as your bot and the basic ones work fine, was wondering how easy it would be to adapt this engine to the same.

https://discord.com/developers/docs/tutorials/hosting-on-cloudflare-workers

I was following this tutorial but the main issue is how the handle requests coming in:

```javascript
export default {
  /**
   * Every request to a worker will start in the `fetch` method.
   * Verify the signature with the request, and dispatch to the router.
   * @param {*} request A Fetch Request object
   * @param {*} env A map of key/value pairs with env vars and secrets from the cloudflare env.
   * @returns
   */
  async fetch(request, env) {
    if (request.method === 'POST') {
      // Using the incoming headers, verify this request actually came from discord.
      const signature = request.headers.get('x-signature-ed25519');
      const timestamp = request.headers.get('x-signature-timestamp');
      console.log(signature, timestamp, env.DISCORD_PUBLIC_KEY);
      const body = await request.clone().arrayBuffer();
      const isValidRequest = verifyKey(
        body,
        signature,
        timestamp,
        env.DISCORD_PUBLIC_KEY
      );
      if (!isValidRequest) {
        console.error('Invalid Request');
        return new Response('Bad request signature.', { status: 401 });
      }
    }

    // Dispatch the request to the appropriate route
    return router.handle(request, env);
  },
};
```
and in their example they are using an `itty-router` instance to handle the requests.

The fetch method will get the requests in and if they are legit it passes to a handler and uses the discord-interactions code right now which is good but a bit verbose.

Any ideas would be appreciated, thank you.

**Comment by monbrey** [comment, marked as answer]
discord.js is designed for gateway connections, generally running as a Node application. It's not intended to be used for http interactions.

However, the discord.js library is composed of several packages that can be used independently - you may be interested in `@discordjs/rest` for making API calls and `@discordjs/builders` for building components, embeds etc in responses.

There's also a standalone structures package being built for class structures to represent Discord objects (User, Message etc)

**Reply by monbrey** [reply]
Correction: `/rest` does not work, but I'd also reference `discord-api-types` to make your TypeScript easier, if you use it

**Reply by imranbarbhuiya** [reply]
or you can use https://github.com/skyra-project/archid-components/tree/main/packages/http-framework too

**Reply by kyranet** [reply]
@imranbarbhuiya that package uses /rest and does not feature worker support at all.

**Reply by null-prophet** [reply]
Thankyou @monbrey I had a go with the command registration part and that was working nicely. 

the code I have written is a bit more verbose (with case statements for handling specific response types etc) but its working.

I started with TS but initially I had a bunch of problems with namespaces etc, mostly in the cloudflare land and putting in some of the feautres I was using from a tutorial. Got it working now with simple JS. If thats stable then great. I'll definitely look at using these APIs in the future. Thank you so much for the reply.

**Reply by Crypt0Beaver** [reply]
Hi,
And for gateway type applications, would Discord.js work on cloudflare workers? They seem compatible with node.js, however when trying juste to import discord.js and starting wrangler there are a bunch of reference errors.

**Reply by NickantX** [reply]
> Hi, And for gateway type applications, would Discord.js work on cloudflare workers? They seem compatible with node.js, however when trying juste to import discord.js and starting wrangler there are a bunch of reference errors.

Cloudflare Workers don't support running a long-lived process, which is exactly what a gateway type application needs to do. The application stays connected to Discord via their websocket gateway and receives events that way. Cloudflare Workers are better suited for modern apps using HTTP Interactions, where Discord sends an HTTP request when an application command is invoked. This does not require a gateway connection, and is a popular way to easily build a scalable Discord app. But discord.js is not designed for these types of applications.

**Reply by sayyidyofa** [reply]
So why does there a "Powered by Workers" badge in the readme

I wasted my time trying to make it work... I'll just use raw api calls then

**Reply by didinele** [reply]
> So why does there a "Powered by Workers" badge in the readme

Because we have an OSS partnership with them, like we do with Vercel.

That being said, you could use `@discordjs/core/http-only` in workers for a higher level of abstraction. See https://discordjs.dev/docs/packages/core/1.0.1

**Reply by leodeslf** [reply]
> Cloudflare Workers don't support running a long-lived process.

Is there any update on this?

(I couldn't find anything new in Cloudflare's docs, maybe you know something else.)

**Reply by evelyn** [reply]
Also interested in this.

**Reply by Qjuh** [reply]
Workers remain workers. If you want persistent connections with the gateway a cloudflare worker is the wrong choice of a tool. That will never change since it‘s the purpose of workers to not be persistent.

**Reply by hacknug** [reply]
What about Durable Objects? https://developers.cloudflare.com/durable-objects/

**Reply by NickantX** [reply]
> What about Durable Objects? https://developers.cloudflare.com/durable-objects/

This is for data storage and state management. Connecting to the gateway works over a WebSocket connection and requires that connection to stay open to receive events from it. This is not a use case Cloudflare Workers are intended for, they are designed to only run when they are invoked, such as from an HTTP request.

**Reply by hacknug** [reply]
Don't they use DO to do exactly that in this tutorial? https://youtu.be/ML1vOUeA-JM?si=rsgrQtlrGjkb8EZo
Also from their docs: https://developers.cloudflare.com/durable-objects/reference/websockets/

**Reply by NickantX** [reply]
> Don't they use DO to do exactly that in this tutorial? https://youtu.be/ML1vOUeA-JM?si=rsgrQtlrGjkb8EZo Also from their docs: https://developers.cloudflare.com/durable-objects/reference/websockets/

Seems that way, although it is yet to be seen how well that would work with Discord bots, and there are no libraries that support it.

**Comment by mia-riezebos** [comment]
until recently it was impossible (i.e. disabled) to connect to the gateway api from Cloudflare Workers, but they seem to have enabled this now.

https://github.com/discord/discord-api-docs/issues/6145#issuecomment-3618601748

**Reply by miguilimzero** [reply]
And with this library still not possible?

**Reply by Amgelo563** [reply]
@miguilimzero Correct, the comments above by the team are up to date. It's unlikely d.js will ever support it since its entire purpose is to make gateway-connected bots. Like the comments above said, you can use `@discordjs/core/http-only` to make rest calls though.
